Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

```diff
@@ -161,30 +119,105 @@ MacroscopicProperties::InitData ()
+    if (m_epsilon_s == "constant") {
+
+        m_eps_mf->setVal(m_epsilon);
+
+    } else if (m_epsilon_s == "parse_epsilon_function") {
+
+        InitializeMacroMultiFabUsingParser(m_eps_mf.get(), m_epsilon_parser->compile<3>(), lev);
+
+    }
+    // Initialize mu
+    if (m_mu_s == "constant") {
+
+        m_mu_mf->setVal(m_mu);
+
+    } else if (m_mu_s == "parse_mu_function") {
+
+        InitializeMacroMultiFabUsingParser(m_mu_mf.get(), m_mu_parser->compile<3>(), lev);
+
+    }
+
+    amrex::IntVect sigma_stag = m_sigma_mf->ixType().toIntVect();
+    amrex::IntVect epsilon_stag = m_eps_mf->ixType().toIntVect();
+    amrex::IntVect mu_stag = m_mu_mf->ixType().toIntVect();
+    amrex::IntVect Ex_stag = warpx.getEfield_fp(0,0).ixType().toIntVect();
+    amrex::IntVect Ey_stag = warpx.getEfield_fp(0,1).ixType().toIntVect();
+    amrex::IntVect Ez_stag = warpx.getEfield_fp(0,2).ixType().toIntVect();
 
     for ( int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+        sigma_IndexType[idim]   = sigma_stag[idim];
+        epsilon_IndexType[idim] = epsilon_stag[idim];
+        mu_IndexType[idim]      = mu_stag[idim];
         Ex_IndexType[idim]      = Ex_stag[idim];
         Ey_IndexType[idim]      = Ey_stag[idim];
         Ez_IndexType[idim]      = Ez_stag[idim];
-        Bx_IndexType[idim]      = Bx_stag[idim];
-        By_IndexType[idim]      = By_stag[idim];
-        Bz_IndexType[idim]      = Bz_stag[idim];
+        macro_cr_ratio[idim]    = 1;
     }
 #if (AMREX_SPACEDIM==2)
+        sigma_IndexType[2]   = 0;
+        epsilon_IndexType[2] = 0;
+        mu_IndexType[2]      = 0;
         Ex_IndexType[2]      = 0;
         Ey_IndexType[2]      = 0;
         Ez_IndexType[2]      = 0;
-        Bx_IndexType[2]      = 0;
-        By_IndexType[2]      = 0;
-        Bz_IndexType[2]      = 0;
+        macro_cr_ratio[2]    = 1;
 #endif
+}
+
+void
+MacroscopicProperties::InitializeMacroMultiFabUsingParser (
+                       amrex::MultiFab *macro_mf,
+                       amrex::ParserExecutor<3> const& macro_parser,
+                       const int lev)
+{
+    auto& warpx = WarpX::GetInstance();
+    const auto dx_lev = warpx.Geom(lev).CellSizeArray();
+    const RealBox& real_box = warpx.Geom(lev).ProbDomain();
+    IntVect iv = macro_mf->ixType().toIntVect();
+    for ( MFIter mfi(*macro_mf, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+        // Initialize ghost cells in addition to valid cells
+
+        const Box& tb = mfi.tilebox( iv, macro_mf->nGrowVect());
+        auto const& macro_fab =  macro_mf->array(mfi);
+        amrex::ParallelFor (tb,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                // Shift x, y, z position based on index type
+                Real fac_x = (1._rt - iv[0]) * dx_lev[0] * 0.5_rt;
+                Real x = i * dx_lev[0] + real_box.lo(0) + fac_x;
+
+                Real fac_y = (1._rt - iv[1]) * dx_lev[1] * 0.5_rt;
+                Real y = j * dx_lev[1] + real_box.lo(1) + fac_y;
```

Comment:
Just to clarify: This `y` label here represents effectively the `z` coordinate in 2D? I'm fine with leaving it like this, just want to make sure that 2D and 3D are treated correctly.

Reply:
oh ! you are right and I fixed this so it works for 2D as well now! Thanks for catching this, @EZoni